taichi/transforms/whole_kernel_cse.cpp

```diff
@@ -95,13 +115,16 @@ class WholeKernelCSE : public BasicStmtVisitor {
   void visit(Stmt *stmt) override {
     if (!stmt->common_statement_eliminable())
       return;
+    // container_statement is no need to be CSE
+    if (stmt->is_container_statement())
+      return;
     // Generic visitor for all CSE-able statements.
     if (is_done(stmt)) {
-      visible_stmts_.back()[std::type_index(typeid(*stmt))].insert(stmt);
+      visible_stmts_.back()[Myhash{}(stmt)].insert(stmt);
       return;
     }
     for (auto &scope : visible_stmts_) {
-      for (auto &prev_stmt : scope[std::type_index(typeid(*stmt))]) {
+      for (auto &prev_stmt : scope[Myhash{}(stmt)]) {
```

Comment:
We can create an instance of `Myhash{}(stmt)` to avoid traversing `stmt`'s operands multiple times.

Also, it seems that using a `struct` here is an overkill -- we can simply replace `std::size_t Myhash::operator()` with something like `static std::size_t operand_type_hash()`.

Reply:
Thanks! I have made the change!